compiler/src/iree/compiler/Dialect/Util/Transforms/IPO.cpp

```diff
@@ -622,9 +622,25 @@ static bool applyCallChanges(FuncAnalysis &analysis,
   // Erase old op now that all uses are (or should be) replaced.
   callOp.erase();
 
+  callOp = newCallOp;
   return true;
 }
 
+// Returns true if |funcOp| performs no work (no args/results, no ops).
+// We could make this a little smarter in the future by checking that there's
+// no side-effecting work.
+static bool isFuncEmpty(FunctionOpInterface funcOp) {
+  if (funcOp.isExternal()) {
+    return false;
+  } else if (funcOp.getNumArguments() > 0 || funcOp.getNumResults() > 0) {
+    return false;
+  } else if (funcOp.getBlocks().size() > 1) {
+    return false;
+  } else {
+    return funcOp.front().getOperations().size() == 1;
+  }
+}
```

Comment:
Could we lean on the standard DCE pass for this instead? Is this an optimization while running the pass itself, and DCE could handle cleanup if it ran after?

Reply:
unfortunately not; local DCE (done during canonicalization) can't remove the calls because it doesn't know what's on the other side of the call, and symbol DCE can't remove the funcs because it only looks as to whether they are used and not their contents (and with the calls still present the symbol is pinned as used).

Reply:
(we could have a pass that annotates functions with nosideeffects-style attrs, then our own `util.call` canonicalizer could drop calls with no results that have nosideeffects, but it's cheaper to do it here and fits the interprocedural nature of the pass)